torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
+    if (chunk_desc.nSubtensors == 1) {
+      try {
+        map_size = at::infer_size(map_size, arg.sizes());
+      } catch (std::exception& e) {
+        return at::nullopt;
+      }
     } else {
-      JIT_ASSERT(canCompileOnCPU());
-      raw_func = new CPUFusionFunction(name, agraph, config_);
+      auto tensor_sizes = arg.sizes().vec();
+      int64_t num_chunks = chunk_desc.nSubtensors;
+      int64_t dim = chunk_desc.dim;
+      if (dim < 0) {
+        dim += arg.dim();
+      }
+      JIT_ASSERT(dim >= 0 && dim <= static_cast<int64_t>(tensor_sizes.size()));
+      if (tensor_sizes[dim] % num_chunks != 0) {
+        return at::nullopt;
+      }
+      tensor_sizes[dim] /= num_chunks;
+      try {
+        map_size = at::infer_size(map_size, tensor_sizes);
+      } catch (std::exception& e) {
+        return at::nullopt;
+      }
     }
-    it = cache.emplace(key_, std::shared_ptr<CompiledFusionFunction>(raw_func)).first;
   }
-  return it->second;
+
+  return {map_size};
 }
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(Node* fusion_group) {
-  auto & graph = *fusion_group->g(attr::Subgraph);
-  AnnotatedGraph agraph(graph, fusion_group->i(attr::device));
-  for(auto & input : graph.inputs()) {
-    auto t = input->type()->expect<TensorType>();
-    agraph.input_desc.emplace_back(t);
+// See Note [Run-time shape checking code] for more explanation on the algorithm.
+at::optional<std::vector<int64_t>> FusedKernelCache::canRunKernel(at::TensorList args) {
+  AT_CHECK(args.size() == input_chunks.size(),
+           "Expected ", input_chunks.size(), " arguments, but got ", args.size());
+
+  at::optional<std::vector<int64_t>> map_size;
+  for (const auto & broadcast_group : input_broadcast_groups) {
+    if (!map_size) {
+      map_size = getMapSize(args, broadcast_group);
+      if (!map_size) {
+        return at::nullopt;
+      }
+    } else {
+      auto group_map_size = getMapSize(args, broadcast_group);
+      // NB: this checks that group_map_size is defined AND equal to map_size
+      if (map_size != group_map_size) {
+        return at::nullopt;
+      }
+    }
   }
-  for(auto & output : graph.outputs()) {
-    auto t = output->type()->expect<TensorType>();
-    agraph.output_desc.emplace_back(t);
+  return map_size;
+}
+
+void FusedKernelCache::runFallback(Stack& stack) {
+  InterpreterState(fallback_code).runOneStage(stack);
+}
+
+void FusedKernelCache::expandArgs(std::vector<at::Tensor>& args, std::vector<int64_t>& map_size) {
+  for (size_t i = 0; i < args.size(); ++i) {
+    auto & arg = args[i];
+    auto & pdesc = input_chunks[i];
+    if (pdesc.nSubtensors == 1) {
+      if (arg.sizes().equals(map_size)) continue;
+      arg = arg.expand(map_size);
+    } else {
+      map_size.at(pdesc.dim) *= pdesc.nSubtensors;
+      if (!arg.sizes().equals(map_size)) {
+        arg = arg.expand(map_size);
+      }
+      map_size.at(pdesc.dim) /= pdesc.nSubtensors;
```

Comment:
Yuck. Understandable, but yuck.

Reply:
I understand, but I really don't want to reallocate this shape for every single chunk that we have...

Reply:
All of the chunk and cat stuff is really messy in the fuser for 2 reasons:
1. we don't have mutable operators in the normal graph, so we cannot express the concat optimization as a rewrite that would instruct the fuser to mutate narrowed tensors.
2. Tensor allocation in chunk and cat is so expensive that slows execution time. The only reasons we do chunk inside the fuser to because we can do chunk faster if we don't need to allocate a tensor. 

We should fix these two things and then remove chunk and cat from the fuser, they are better handled elsewhere. 

In the meantime we might consider making a smaller tensor holder specific to the fuser that can do fast narrows. Then chunk and cat could be handled on the periphery of the fuser, rather than handling the PartitionDesc objects in all of these places.  Not for this PR, but something that we should consider as a refactor before doing major further feature additions.